File: app/src/main/java/com/przemolab/oknotifier/asyncTasks/SqliteContestLoader.java
package com.przemolab.oknotifier.asyncTasks;

import android.content.Context;
import android.support.v4.content.AsyncTaskLoader;

import com.przemolab.oknotifier.enums.SortOrder;
import com.przemolab.oknotifier.models.Contest;
import com.przemolab.oknotifier.modules.NotifierRepository;

import java.util.List;

import timber.log.Timber;

public class SqliteContestLoader extends AsyncTaskLoader<List<Contest>> {

    private List<Contest> contests = null;
    private NotifierRepository notifierRepository;
    private SortOrder sortOrder;

    public SqliteContestLoader(Context context, NotifierRepository notifierRepository, SortOrder sortOrder) {
        super(context);
        this.notifierRepository = notifierRepository;
        this.sortOrder = sortOrder;
    }

    @Override
    public List<Contest> loadInBackground() {
        try {
            Timber.d("Loading contests from SQLite");
            return notifierRepository.getAll(sortOrder);
        } catch (Exception ex) {
            Timber.e(ex);
            return null;
        }
    }

    @Override
    protected void onStartLoading() {
        if (contests == null) {
            forceLoad();
        } else {
            deliverResult(contests);
        }
    }

    @Override
    public void deliverResult(List<Contest> data) {
        contests = data;
        super.deliverResult(data);
    }
}

File: app/src/main/java/com/przemolab/oknotifier/sync/SqliteContestLoader.kt
package com.przemolab.oknotifier.sync

import android.content.Context
import android.support.v4.content.AsyncTaskLoader
import com.przemolab.oknotifier.data.ContestEntry

import com.przemolab.oknotifier.enums.SortOrder
import com.przemolab.oknotifier.interfaces.INotifierRepository

import timber.log.Timber

class SqliteContestLoader(context: Context,
                          private val notifierRepository: INotifierRepository,
                          private val sortOrder: SortOrder) : AsyncTaskLoader<List<ContestEntry>>(context) {

    private var contests: List<ContestEntry>? = null

    override fun loadInBackground(): List<ContestEntry>? {
        return try {
            Timber.d("Loading contests from SQLite")
            notifierRepository.getAll(sortOrder)
        } catch (ex: Exception) {
            Timber.e(ex)
            null
        }
    }

    override fun onStartLoading() {
        if (contests == null) {
            forceLoad()
        } else {
            deliverResult(contests)
        }
    }

    override fun deliverResult(data: List<ContestEntry>?) {
        contests = data
        super.deliverResult(data)
    }
}
